text/3371-cargo-target-directories.md

```diff
@@ -0,0 +1,328 @@
+bringing the registry cache in their build directory for example.
+
+## Stabilize the naming scheme
+
+I feel this require an hard-to-break naming scheme (a recent hash algorithm should be good enough in 99% of the cases but collisions are always possible), which I don't have the skills nor motivation to design. Instead, I prefer explicitely telling the naming scheme is not to be considered stable and allow more invested people to experiment with the feature and find something solid.
+
+What's more, by explicitely not stabilizing it (and maybe voluntarily changing it between versions sometimes, since a version change recompiles everything anyway ?) we can instead reroute people and tools towards `CARGO_TARGET_DIR` / `cargo metadata` instead, which are much more likely to be suited to their use case if they need the path to the target directory.
+
+## Use only the hash, not the name of the workspace directory
+
+While this solution is just as easy to work with for tools, I think having a somewhat human-readable part in the name make it easier to work with through things like logging or just listing the subdirectories in `CARGO_TARGET_DIRECTORIES`. It also makes the path printed by `cargo run` or
+the local URL used when calling `cargo doc --open` much clearer.
+
+Ultimately, tools are not negatively affected since they should be using `CARGO_TARGET_DIR` or `cargo metadata` and humans have an easier time working with such paths so I think it's worthwhile to include the workspace's name.
+
+## Just use `targo`
+
+While a very nice tool, `targo` is not integrated with `cargo` and has a few shortcomings:
+
+- It uses symlinks, which are not always handled well by other tools. Specifically, since it's not integrated inside `cargo`, it uses a `target` symlink to avoid having to remap `cargo`'s execution using `CARGO_TARGET_DIR` and such,making it less useful for external build tools that would use this functionality. Using such a symlink also means `cargo clean` does not work, it just removes the symlink and not the data.
+- It completely ignores `CARGO_TARGET_DIR`-related options, which again may break workflows.
+- It needs more metadata to work well, which means an external tool using it would have to understand that metadata too.
+- It uses `$CARGO_HOME/targo` to place its cache, making it less useful for external build tools and people wanting to separate caches and configuration.
+- It needs to intercept `cargo`'s arguments, making it more brittle than an integrated solution.
+
+Some of those could be fixed of course, and I don't expect `cargo`'s `--target-dir` and `--manifest-path` to change or disappear anytime soon, but still, it could happen. An external tool like `targo` will never be able to
+solve some of these or ensure forward compatibility as well as the solution proposed in this RFC.
+
+On the other hand, `targo` is already here and working for at least one person, making it the most viable alternative for now.
```

Comment:
Could you also compare the caching strategy with targo, not just the use of an external tool?

Reply:
(maybe also link to targo for easier reference)

Reply:
For example, targo provides backlinks.  Is there a reason for us to do that or not?

Reply:
done

Reply:
I'm not seeing an examination of targo's caching strategy and how it compares

Reply:
There is no caching strategy beyond what cargo does inside `target/`, `targo` is basically only a wrapper to automatically make the `target/` dir a symlink to something else.

Reply:
done